controllers/controller/workspacerouting/workspacerouting_controller.go

```diff
@@ -14,33 +14,49 @@ package workspacerouting
 
 import (
 	"context"
+	"errors"
 	"fmt"
 
 	"github.com/devfile/devworkspace-operator/controllers/controller/workspacerouting/solvers"
+	maputils "github.com/devfile/devworkspace-operator/internal/map"
 	"github.com/devfile/devworkspace-operator/pkg/config"
 	"github.com/google/go-cmp/cmp"
+
+	"github.com/go-logr/logr"
 	routeV1 "github.com/openshift/api/route/v1"
 	corev1 "k8s.io/api/core/v1"
 	"k8s.io/api/extensions/v1beta1"
-	"k8s.io/apimachinery/pkg/api/errors"
-	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
-	"sigs.k8s.io/controller-runtime/pkg/reconcile"
-
-	"github.com/go-logr/logr"
+	k8sErrors "k8s.io/apimachinery/pkg/api/errors"
 	"k8s.io/apimachinery/pkg/runtime"
 	ctrl "sigs.k8s.io/controller-runtime"
 	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
 
 	controllerv1alpha1 "github.com/devfile/devworkspace-operator/apis/controller/v1alpha1"
 )
 
+var (
+	RoutingNotSupported = errors.New("routingclass not supported by this controller")
+	NoSolversEnabled    = errors.New("reconciler does not define GetSolverFunc")
+)
+
 const workspaceRoutingFinalizer = "workspacerouting.controller.devfile.io"
 
+// WorkspaceRoutingSolverFunc is a function that obtains a Solver
+// (see github.com/devfile/devworkspace-operator/controllers/controller/workspacerouting/solvers)
+// for a particular WorkspaceRouting instance. This function should return an RoutingNotSupported if
+// the routingClass is not recognized, and any other error if the routingClass is invalid (e.g. an OpenShift-only
+// routingClass on a vanilla Kubernetes platform).
+type WorkspaceRoutingSolverFunc func(routingClass controllerv1alpha1.WorkspaceRoutingClass) (solver solvers.RoutingSolver, err error)
```

Comment:
Would it make sense to pass a `Client` to this function as well so that the solvers can use it for querying/manipulating the cluster or is it ok for the solvers to obtain the client by other means?

Reply:
I assume passing the client could make sense if we also want to reuse the main reconciliation loop code. wdyt @amisevsk ?

Reply:
We could pass the client, but I don't see why it should be needed; this function is basically a mapping from `routingClass -> Solver`. Is there a reason we would need to reach out to the cluster here?

The Solver interface in general does not include any cluster-related actions -- it simply returns a list of objects that we want to exist on the cluster. The main reconcile loop is responsible for making what's on the cluster match what's returned by the solver.

Reply:
Thinking about specific use-cases, do we foresee needing to extend the `RoutingObjects` struct? For a given workspace, the workspacerouting reconciler will make sure the services/routes/ingresses returned by `GetSpecObjects` exist on the cluster, and any other objects are removed. 

We can discuss this for a separate PR as the need becomes clearer.

Reply:
Of course, the detail I forgot here is that `GetSolverFunc` has to work in the predicates as well, to properly filter events from triggering reconciles. I've removed the change from the PR; we can work to add it later if we find it necessary.

Reply:
it's going to be breaking change, which should be harder to do later, right?

Also, I wonder if returning an object that is not used in predicate influences performance?

Maybe include RoutingSolverFactory (probably wrong naming for go):
```
  isSupported()
  getSolver() <-  currently Get Solvers returns a new instance... even though it's not used at all.
```